File: src/projects/midterm/load-obj.ts
export type Vertex = [x: number, y: number, z: number];
export type Face = Vertex[];

export const load_obj = (obj: string) => {
  const lines = obj.split("\n");
  const vertices: Vertex[] = [];
  const faces: Face[] = [];
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const line_num = i + 1;
    if (line.startsWith("#") || line.trim() === "") continue;
    const chunks = line.split(/\s+/g);
    const command = chunks[0];
    if (
      command === "mtllib" ||
      command === "o" ||
      command === "vn" ||
      command === "vt" ||
      command === "s"
    ) {
      // ignore/no need to handle (at least for now)
    } else if (command === "v") {
      // @ts-expect-error
      vertices.push([...chunks.slice(1, 4).map((chunk) => Number(chunk)), 1]);
    } else if (command === "f") {
      const face_vertices = chunks.slice(1);
      if (face_vertices.length !== 3)
        throw new Error(
          "Only triangles are supported in loading obj files for now",
        );
      faces.push(
        face_vertices.map((chunk) => {
          // eslint-disable-next-line @typescript-eslint/naming-convention
          const [vertex_index, _texture_index, _normal_index] = chunk
            .split("/")
            .map((c) => Number(c));
          return vertices[vertex_index - 1];
        }),
      );
    } else {
      throw new Error(`Unrecognized command on line ${line_num}: ${command}`);
    }
  }
  return faces;
};

File: src/projects/midterm/load_obj.rs
use nalgebra::Point3;

use crate::face::Face;

pub(crate) fn load_obj(obj: &str) -> Vec<Face<f64>> {
    let lines = obj.split('\n');
    let mut vertices: Vec<Point3<f64>> = vec![];
    let mut faces: Vec<Face<f64>> = vec![];
    for (i, line) in lines.enumerate() {
        let line_num = i + 1;
        if line.starts_with('#') || line.trim() == "" {
            continue;
        };
        let chunks: Vec<&str> = line.split_whitespace().collect();
        let command = chunks[0];
        if command == "mtllib"
            || command == "o"
            || command == "vn"
            || command == "vt"
            || command == "s"
        {
            // ignore/no need to handle (at least for now)
        } else if command == "v" {
            vertices.push(Point3::new(
                chunks[1].parse().unwrap(),
                chunks[2].parse().unwrap(),
                chunks[3].parse().unwrap(),
            ));
        } else if command == "f" {
            let face_vertices = &chunks[1..];
            if face_vertices.len() != 3 {
                panic!("Only triangles are supported in loading obj files for now",);
            }
            let face_points: Vec<Point3<f64>> = face_vertices
                .iter()
                .map(|face_vertex| {
                    let vertex_id: usize = face_vertex.split('/').next().unwrap().parse().unwrap();
                    let vertex_id = vertex_id - 1;
                    vertices[vertex_id]
                })
                .collect();
            let new_face = Face::new(face_points);
            faces.push(new_face);
        } else {
            panic!("Unrecognized command on line {}: {}", line_num, command);
        }
    }
    faces
}
